File: app/src/main/java/com/quran/labs/androidquran/ui/util/ImageAyahUtils.java
package com.quran.labs.androidquran.ui.util;

import android.graphics.Matrix;
import android.graphics.RectF;
import android.support.annotation.NonNull;
import android.util.SparseArray;
import android.widget.ImageView;

import com.quran.labs.androidquran.common.AyahBounds;
import com.quran.labs.androidquran.data.SuraAyah;
import com.quran.labs.androidquran.widgets.AyahToolBar;
import com.quran.labs.androidquran.widgets.HighlightingImageView;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import timber.log.Timber;

public class ImageAyahUtils {

   private static SuraAyah getAyahFromKey(String key){
      String[] parts = key.split(":");
     SuraAyah result = null;
      if (parts.length == 2){
         try {
            int sura = Integer.parseInt(parts[0]);
            int ayah = Integer.parseInt(parts[1]);
            result = new SuraAyah(sura, ayah);
         }
         catch (Exception e){
           // no op
         }
      }
      return result;
   }

   public static SuraAyah getAyahFromCoordinates(
           Map<String, List<AyahBounds>> coords,
           HighlightingImageView imageView, float xc, float yc) {
      if (coords == null || imageView == null){ return null; }

      float[] pageXY = getPageXY(xc, yc, imageView);
      if (pageXY == null){ return null; }
      float x = pageXY[0];
      float y = pageXY[1];

      int closestLine = -1;
      int closestDelta = -1;

      final SparseArray<List<String>> lineAyahs = new SparseArray<>();
      final Set<String> keys = coords.keySet();
      for (String key : keys){
         List<AyahBounds> bounds = coords.get(key);
         if (bounds == null){ continue; }

         for (AyahBounds b : bounds){
            // only one AyahBound will exist for an ayah on a particular line
            int line = b.getLine();
            List<String> items = lineAyahs.get(line);
            if (items == null){
               items = new ArrayList<>();
            }
            items.add(key);
            lineAyahs.put(line, items);

            final RectF boundsRect = b.getBounds();
            if (boundsRect.contains(x, y)) {
               return getAyahFromKey(key);
            }

            int delta = Math.min((int) Math.abs(boundsRect.bottom - y),
                    (int) Math.abs(boundsRect.top - y));
            if (closestDelta == -1 || delta < closestDelta){
               closestLine = b.getLine();
               closestDelta = delta;
            }
         }
      }

      if (closestLine > -1){
         int leastDeltaX = -1;
         String closestAyah = null;
         List<String> ayat = lineAyahs.get(closestLine);
         if (ayat != null){
            Timber.d("no exact match, %d candidates.", ayat.size());
            for (String ayah : ayat){
               List<AyahBounds> bounds = coords.get(ayah);
               if (bounds == null){ continue; }
               for (AyahBounds b : bounds){
                  if (b.getLine() > closestLine){
                     // this is the last ayah in ayat list
                     break;
                  }

                  final RectF boundsRect = b.getBounds();
                  if (b.getLine() == closestLine){
                     // if x is within the x of this ayah, that's our answer
                     if (boundsRect.right >= x && boundsRect.left <= x){
                        return getAyahFromKey(ayah);
                     }

                     // otherwise, keep track of the least delta and return it
                     int delta = Math.min((int) Math.abs(boundsRect.right - x),
                             (int) Math.abs(boundsRect.left - x));
                     if (leastDeltaX == -1 || delta < leastDeltaX){
                        closestAyah = ayah;
                        leastDeltaX = delta;
                     }
                  }
               }
            }
         }

         if (closestAyah != null){
            Timber.d("fell back to closest ayah of %s", closestAyah);
            return getAyahFromKey(closestAyah);
         }
      }
      return null;
   }

  public static AyahToolBar.AyahToolBarPosition getToolBarPosition(
      @NonNull List<AyahBounds> bounds, @NonNull Matrix matrix,
      int screenWidth, int screenHeight, int toolBarWidth, int toolBarHeight) {
    boolean isToolBarUnderAyah = false;
    AyahToolBar.AyahToolBarPosition result = null;
    final int size = bounds.size();

    RectF chosenRect;
    if (size > 0) {
      RectF firstRect = new RectF();
      AyahBounds chosen = bounds.get(0);
      matrix.mapRect(firstRect, chosen.getBounds());
      chosenRect = new RectF(firstRect);

      float y = firstRect.top - toolBarHeight;
      if (y < toolBarHeight) {
        // too close to the top, let's move to the bottom
        chosen = bounds.get(size - 1);
        matrix.mapRect(chosenRect, chosen.getBounds());
        y = chosenRect.bottom;
        if (y > (screenHeight - toolBarHeight)) {
          y = firstRect.bottom;
          chosenRect = firstRect;
        }
        isToolBarUnderAyah = true;
      }

      final float midpoint = chosenRect.centerX();
      float x = midpoint - (toolBarWidth / 2);
      if (x < 0 || x + toolBarWidth > screenWidth) {
        x = chosenRect.left;
        if (x + toolBarWidth > screenWidth) {
          x = screenWidth - toolBarWidth;
        }
      }

      result = new AyahToolBar.AyahToolBarPosition();
      result.x = x;
      result.y = y;
      result.pipOffset = midpoint - x;
      result.pipPosition = isToolBarUnderAyah ?
          AyahToolBar.PipPosition.UP : AyahToolBar.PipPosition.DOWN;
    }
    return result;
  }

  private static float[] getPageXY(
      float screenX, float screenY, ImageView imageView) {
    if (imageView.getDrawable() == null) {
      return null;
    }

    float[] results = null;
    Matrix inverse = new Matrix();
    if (imageView.getImageMatrix().invert(inverse)) {
      results = new float[2];
      inverse.mapPoints(results, new float[]{screenX, screenY});
    }
    return results;
  }

  public static RectF getYBoundsForHighlight(
      Map<String, List<AyahBounds>> coordinateData, int sura, int ayah) {
    final List<AyahBounds> ayahBounds = coordinateData.get(sura + ":" + ayah);
    if (ayahBounds == null) {
      return null;
    }

    RectF ayahBoundsRect = null;
    for (AyahBounds bounds : ayahBounds) {
      if (ayahBoundsRect == null) {
        ayahBoundsRect = bounds.getBounds();
      } else {
        ayahBoundsRect.union(bounds.getBounds());
      }
    }
    return ayahBoundsRect;
  }
}

File: app/src/main/java/com/quran/labs/androidquran/ui/util/ImageAyahUtils.kt
package com.quran.labs.androidquran.ui.util

import android.graphics.Matrix
import android.graphics.RectF
import android.util.SparseArray
import android.widget.ImageView
import com.quran.data.model.SuraAyah
import com.quran.data.model.selection.SelectionIndicator
import com.quran.data.model.selection.SelectionRectangle
import com.quran.labs.androidquran.view.HighlightingImageView
import com.quran.page.common.data.AyahBounds
import timber.log.Timber
import kotlin.math.abs
import kotlin.math.min

object ImageAyahUtils {
  private fun getAyahFromKey(key: String): SuraAyah? {
    val parts = key.split(":")
    val correctNumberOfParts = parts.size == 2
    val sura = if (correctNumberOfParts) parts[0].toIntOrNull() else null
    val ayah = if (correctNumberOfParts) parts[1].toIntOrNull() else null
    return if (sura != null && ayah != null) {
      SuraAyah(sura, ayah)
    } else {
      null
    }
  }

  fun getAyahFromCoordinates(coords: Map<String, List<AyahBounds>>?,
                             imageView: HighlightingImageView?, xc: Float, yc: Float): SuraAyah? {
    return getAyahBoundsFromCoordinates(coords, imageView, xc, yc)?.first
  }

  private fun getAyahBoundsFromCoordinates(
      coords: Map<String, List<AyahBounds>>?,
      imageView: HighlightingImageView?,
      xc: Float,
      yc: Float
  ): Pair<SuraAyah?, AyahBounds>? {
    if (coords == null || imageView == null) {
      return null
    }

    val pageXY = getPageXY(xc, yc, imageView) ?: return null

    val x = pageXY[0]
    val y = pageXY[1]
    var closestLine = -1
    var closestDelta = -1

    val lineAyahs = SparseArray<MutableList<String>>()
    val keys = coords.keys
    for (key in keys) {
      val bounds = coords[key] ?: continue
      for (b in bounds) {
        // only one AyahBound will exist for an ayah on a particular line
        val line = b.line
        var items = lineAyahs[line]
        if (items == null) {
          items = ArrayList()
        }
        items.add(key)
        lineAyahs.put(line, items)
        val boundsRect = b.bounds
        if (boundsRect.contains(x, y)) {
          return Pair(getAyahFromKey(key), b)
        }
        val delta = min(
          abs(boundsRect.bottom - y).toInt(),
          abs(boundsRect.top - y).toInt()
        )
        if (closestDelta == -1 || delta < closestDelta) {
          closestLine = b.line
          closestDelta = delta
        }
      }
    }

    if (closestLine > -1) {
      var leastDeltaX = -1
      var closestAyah: String? = null
      var closestAyahBounds: AyahBounds? = null
      val ayat: List<String>? = lineAyahs[closestLine]
      if (ayat != null) {
        Timber.d("no exact match, %d candidates.", ayat.size)
        for (ayah in ayat) {
          val bounds = coords[ayah] ?: continue
          for (b in bounds) {
            if (b.line > closestLine) {
              // this is the last ayah in ayat list
              break
            }
            val boundsRect = b.bounds
            if (b.line == closestLine) {
              // if x is within the x of this ayah, that's our answer
              if (boundsRect.right >= x && boundsRect.left <= x) {
                return Pair(getAyahFromKey(ayah), b)
              }

              // otherwise, keep track of the least delta and return it
              val delta = min(
                abs(boundsRect.right - x).toInt(),
                abs(boundsRect.left - x).toInt()
              )
              if (leastDeltaX == -1 || delta < leastDeltaX) {
                closestAyah = ayah
                closestAyahBounds = b
                leastDeltaX = delta
              }
            }
          }
        }
      }

      if (closestAyah != null && closestAyahBounds != null) {
        Timber.d("fell back to closest ayah of %s", closestAyah)
        return Pair(getAyahFromKey(closestAyah), closestAyahBounds)
      }
    }
    return null
  }

  fun getToolBarPosition(
    bounds: List<AyahBounds>,
    matrix: Matrix,
    xPadding: Int,
    yPadding: Int
  ): SelectionIndicator {

    return if (bounds.isNotEmpty()) {
      val first = bounds.first()
      val last = bounds.last()

      val mappedRect = RectF()
      matrix.mapRect(mappedRect, first.bounds)
      val top = SelectionRectangle(
        mappedRect.left + xPadding,
        mappedRect.top + yPadding,
        mappedRect.right + xPadding,
        mappedRect.bottom + yPadding
      )
      val bottom = if (first === last) { top }
      else {
        matrix.mapRect(mappedRect, last.bounds)
        SelectionRectangle(
          mappedRect.left + xPadding,
          mappedRect.top + yPadding,
          mappedRect.right + xPadding,
          mappedRect.bottom + yPadding
        )
      }
      SelectionIndicator.SelectedItemPosition(top, bottom)
    } else {
      SelectionIndicator.None
    }
  }

  fun getPageXY(
    screenX: Float, screenY: Float, imageView: ImageView
  ): FloatArray? {
    if (imageView.drawable == null) {
      return null
    }
    var results: FloatArray? = null
    val inverse = Matrix()
    if (imageView.imageMatrix.invert(inverse)) {
      results = FloatArray(2)
      inverse.mapPoints(results, floatArrayOf(screenX, screenY))
      results[1] = results[1] - imageView.paddingTop
    }
    return results
  }

  fun getYBoundsForHighlight(
    coordinateData: Map<String, List<AyahBounds>?>, sura: Int, ayah: Int
  ): RectF? {
    val ayahBounds = coordinateData["$sura:$ayah"] ?: return null
    var ayahBoundsRect: RectF? = null
    for (bounds in ayahBounds) {
      if (ayahBoundsRect == null) {
        ayahBoundsRect = bounds.bounds
      } else {
        ayahBoundsRect.union(bounds.bounds)
      }
    }
    return ayahBoundsRect
  }
}
